### PR TITLE
Validate that rechunking is possible earlier in graph generation

### PR DESCRIPTION
**Pull request by hendrikmakait**
This PR consolidates (almost) all validation and moves it to an earlier point in graph generation to avoid duplication between task-based and P2P-based rechunking (https://github.com/dask/distributed/pull/7856/files#r1205195557)

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: dask/array/rechunk.py [hunk 1]
```diff
@@ -246,6 +238,33 @@ def intersect_chunks(old_chunks, new_chunks):
...
+    old_shapes = tuple(map(sum, old_chunks))
+    new_shapes = tuple(map(sum, new_chunks))
+
+    for old_shape, old_dim, new_shape, new_dim in zip(
+        old_shapes, old_chunks, new_shapes, new_chunks
+    ):
+        if old_shape != new_shape:
+            if not (
+                math.isnan(old_shape) and math.isnan(new_shape)
+            ) or not np.array_equal(old_dim, new_dim, equal_nan=True):
```
Comment (hendrikmakait) [inline dask/array/rechunk.py:261]: Using `np.array_equal` avoids issues with the provenance of nan values (https://github.com/dask/dask/pull/10157#discussion_r1203663937, https://github.com/dask/distributed/pull/7856/files#r1205199004)

File: dask/array/tests/test_rechunk.py [hunk 2]
```diff
@@ -772,13 +772,6 @@ def test_old_to_new_large():
     assert result == expected
 
 
-def test_changing_raises():
-    with pytest.raises(ValueError) as record:
-        old_to_new(((np.nan, np.nan), (4, 4)), ((np.nan, np.nan, np.nan), (4, 4)))
-
-    assert "unchanging" in str(record.value)
```
Comment (wence-) [inline dask/array/tests/test_rechunk.py:779]: Is this test now no longer valid?

File: dask/array/rechunk.py (hunk 1 above)
Comment (wence-) [reply dask/array/rechunk.py:261]: Makes sense, I think.

**Review by wence- — COMMENTED** [review]
Looks overall fine, though I am confused why the test disappears.

File: dask/array/tests/test_rechunk.py (hunk 2 above)
Comment (hendrikmakait) [reply dask/array/tests/test_rechunk.py:779]: Yes, the test is no longer valid. It triggered validation within `old_to_new`. but with this change, `old_to_new` assumes that validation has already taken place (see changes to docstring).

Comment (hendrikmakait) [reply dask/array/tests/test_rechunk.py:779]: I've added the same test case in a different place where it's still valid.

Comment (wence-) [reply dask/array/tests/test_rechunk.py:779]: Thanks, missed that subtlety.

**Review by wence- — APPROVED** [review]

**Review by milesgranger — APPROVED** [review]
Looks good as far as I can see. :+1:

File: dask/array/rechunk.py [hunk 3]
```diff
@@ -160,10 +160,13 @@ def old_to_new(old_chunks, new_chunks):
...
     is unchanged.
 
+    Notes
+    -----
     This function expects that the arguments have been pre-processed by
     :func:`dask.array.core.normalize_chunks`. In particular any ``nan`` values should
     have been replaced (and are so by :func:`dask.array.core.normalize_chunks`)
-    by the canonical ``np.nan``.
+    by the canonical ``np.nan``. It also expects that the arguments have been validated
+    with `_validate_rechunk` and rechunking is thus possible.
```
Comment (jrbourbeau) [inline dask/array/rechunk.py:169]: Nit: we want double backticks around ``_validate_rechunk`` for sphinx

Comment (hendrikmakait) [reply dask/array/rechunk.py:169]: Is there a good way to lint this? 🤔

Comment (jrbourbeau) [reply dask/array/rechunk.py:169]: Not that I've run into (though I've also not specifically looked for one). Though I'd definitely use it if one existed!

**Pull request merged**